### ISSUE TITLE
Proximity BM25 ranking

### ISSUE THREAD

**Issue by kshnurov** [elastic/elasticsearch#2867]
Since we have BM25 similarity in 0.90.0.Beta1, is there any way to implement Sphinx-like Proximity BM25 ranking?

    SPH_RANK_PROXIMITY_BM25, the default ranking mode that uses and combines both phrase proximity and BM25 ranking.

**Comment by vineelyalamarthy** [comment]
@clintongormley   is this some thing to be fixed?  How can I assign this bug to myself ? I m extremely interested in this.

**Comment by kshnurov** [comment]
@kimchy, @s1monw, please add (Phrase proximity + BM25) ranking!
While ES is very good at search, it is very bad at text search **ranking**.

Let's say we have 2 documents:
1. `(2000 words) ... child elephant from America was ... (2000 words)`
2. `Egyptian child ... (2000 words) ... american coffee with elephant ... (2000 words) ... toy elephant`

We're heard something interesting about child elephant from America and now we want to find it: `"match": { "text": "child elephant america"}`. Guess what? 2nd document will be ranked higher(!) because it has 2 'elephants' vs 1 and this is how TF/IDF and current algorithm works.

Things get even worse when you're trying to find a restaurant called "China Street", since you have words "china" and "street" found in hundreds of restaurants. You can add a boost for name field, but then you won't be able to find "roosevelt street restaurants", since restaurants "Roosevelt" and "Lonely Street" will be ranked higher because of the boost. You'll add a boost for address and bang! - now you can't find a restaurant with words "good china restaurant, previously located at roosevelt street" in description. It will be ranked much, much lower than you expect (second page at least) because of the boosts and restaurants "China Street" and "Roosevelt", all china restaurants at Roosevelt st. and etc. This is where phrase proximity is needed to make ranking a lot better.

**Comment by vineelyalamarthy** [comment]
@kimchy @kshnurov  can I work on this bug or some one already took it ? I would like to know where to get started.

**Comment by vineelyalamarthy** [comment]
@kimchy @kshnurov @kshnurov @clintongormley @rmuir This bug sounds very interesting, I want to work on it,you can review my code before check-in , But could you please tell me, how to get started.

**Comment by brusic** [comment]
You still have phrase matching with slop at your disposal, so there is
nothing that is preventing from rolling your own. You can also disable term
frequencies, use bigrams, or all of the above. Many of us have tuned
relevancy with the current tool set. Your queries will get complicated and
the tuning process might take a bit.

Isn't BM25 about document length normalization? Need to look at the sphinx
algorithm to see why proximity is used only there.

As far as contributing to the code, got can find guides online on how to
contribute to a github project. Create a new branch, check in the code and
issue a pull request.  Beware that elasticsearch is notoroius for not
accepting external pull requests. Perhaps that has changed lately (I gave
up).

**Comment by kshnurov** [comment]
@vineelyalamarthy this IS NOT a bug, it's just [how default ES ranking works](https://www.elastic.co/guide/en/elasticsearch/guide/current/scoring-theory.html). In most cases this is more than enough.
@brusic phrase matching is a query type, not a ranking method. Ranking will be the same for the example I've posted. For now ES have [only 2 ranking methods](https://www.elastic.co/guide/en/elasticsearch/guide/current/pluggable-similarites.html) - default one and BM25.

**Comment by brusic** [comment]
All queries contribute to the score, so they implicitly form part of the
ranking method.

I use a complicated set of queries which includes phrase queries to provide
a boost to the score if there is a match. It would be nice to issue a
single query, but alas that is not possible if you want to fine tune
relevancy.

BM25 deals with field length normalization, so it still is TFIDF under the
hood. If your corpus has large variances in the length of a field and term
frequencies are still important, BM25 might be a good approach. It has
nothing to do with proximity.

Agree that this is not a bug, but a feature request. Since there is a
published algorithm, it would be great to incorporate into elasticsearch. I
would give it a try, but I do not have time.

**Comment by vineelyalamarthy** [comment]
@brusic  This will be my first contribution to Elastic Search code base. I would like to work on this , if its approved as a feature request.I have the code set-up in Eclipse. Are there any other steps required to get started.If you know, how to start on this particular task, please let me know.

**Comment by brusic** [comment]
Venkata, it is no different than any other git open source project. Clone
the repo, create a new branch, commit your changes (make sure you squash
your commits into a single one), and then issue a pull request. Add many
tests.

Befriend elastic devs on irc and the mailing list so that you can give some
visibility to your change.

This feature would be great to have. It feels like most do not use
elasticsearch for searching! Tuning is a skill.

**Comment by brusic** [comment]
I should have added that Elasticsearch does not have different scoring
models itself, but simply exposes those found in the Lucene library.
Chances are that any changes you may want to add really belong at the
Lucene layer.

That said, there are many smart people that are committers in Lucene, many
of them are now employed by Elastic. If they have not already implemented
something similar already, there might be a reason. Perhaps someone like
@mikemccand can chime in.

There are other similarity models in Lucene/Elasticsearch, but none of them
will help with this issue.

**Comment by brusic** [comment]
Finally read the Sphinx website and it seems like it does not do anything special. SPH_RANK_PROXIMITY_BM25 is just a scored phrase query while using the scaled TF values calculated by BM25. The two concepts are orthogonal.

Also judging by my quick read, Lucene's sloppy phrase freq scoring is more detailed and granular than Sphinx's longest common sub-sequence (LCS).

**Comment by clintongormley** [comment]
thanks for investigating @brusic - I'm going to close this one

**Comment by brusic** [comment]
Someone please correct me if I am wrong. I haven't even installed Sphinx, but that is my take on things.

**Comment by rmuir** [comment]
Yes, in general here such hacks to BM25 are usually defined as only working on the highest portion of the top-N results[1][2].

So I think any work around this should be done as a rescorer, not a lucene similarity. Today you can already use a `query` rescorer with a sloppy phrase query to do this, see the examples in the documentation[3]. 

I do think it would be good to have rescorer implementations with maybe a better formula than sloppy phrase query, but that is also not obvious since there are still so many varying approaches out there, and all seem kinda ad-hoc. Personally, I don't see any winners, especially since BM25 has term independence assumptions baked in...

1. http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.174.8359
2. http://research.microsoft.com/pubs/144542/ppm.pdf
3. https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-rescore.html

**Comment by kshnurov** [comment]
@brusic you're wrong, Lucene's sloppy phrase freq scoring works *only* for sloppy phrase and span queries. Scoring we're talking about would require putting `n! (n = words count)` or ever more different phrase queries into 1 request.

@clintongormley please reopen this issue. What we need is a rescorer, as @rmuir said, that will add LCS or any similar algorithm on top of the default ranking. It's very simple and probably could be done with scripting, but script won't be fast at all.

**Comment by kshnurov** [comment]
Once again, here's the algorithm. We already have `doc_bm25` value in ES, all we need is to calculate and add `doc_phrase_weight`.
```
doc_phrase_weight = 0
foreach ( field in matching_fields )
{
   field_phrase_weight = max_common_subsequence_length ( query, field )
   doc_phrase_weight += user_weight ( field ) * field_phrase_weight
}

weight = doc_phrase_weight*1000 + integer(doc_bm25*999)
```

**Comment by kshnurov** [comment]
And the LCS algorithm could be found here, it's also quite simple, you just need to add it into ES: https://en.wikibooks.org/wiki/Algorithm_Implementation/Strings/Longest_common_subsequence#Java

**Comment by rmuir** [comment]
> Once again, here's the algorithm.

Hold your horses, where are IR tests for "your algorithm". By default, I'm telling you this algorithm sucks, unless you can prove otherwise. Don't be offended: this is how it works. You have to defend it with some results, otherwise we should not waste our time.

**Comment by kshnurov** [comment]
@rmuir have you at all read this issue before coming and saying 'it sucks'? There's a link to a solid report and tests in [2nd comment](https://github.com/elastic/elasticsearch/issues/2867#issuecomment-16013137): http://stefan.buettcher.org/papers/buettcher_2006_term_proximity.pdf

You can find a lot more in 10 seconds, if all of this serious reports from well-known companies and universities (and the fact that it's been a default ranking method in Sphinx for years) isn't enough for you:
http://www.researchgate.net/publication/225174089_Term_Proximity_Scoring_for_Keyword-Based_Retrieval_Systems
http://research.microsoft.com/pubs/144542/ppm.pdf
http://sifaka.cs.uiuc.edu/czhai/pub/sigir07-prox.pdf
http://ir.dcs.gla.ac.uk/~ronanc/papers/cumminsSIGIR09.pdf
http://people.mpi-inf.mpg.de/~mtb/pub/proximity-spire07.pdf

**Comment by rmuir** [comment]
None of that is good enough for me, its a bogus argument: you link to a bunch of completely different algorithms, different than what you propose.

**Comment by kshnurov** [comment]
@rmuir saying "it sucks" and "is not good enough for me" without saying what is enough and doesn't sucks won't make it move. These links provide more than enough info about why term proximity is important and a lot of algorithms to choose from (with proof and tests, as you requested). 

Sounds like you just don't wanna do anything, since you didn't even spent enough time on reading this links. Saying about Sphinx being a 'bogus argument' is completely weird. @s1monw @clintongormley could you please join and make this conversation constructive, not a "I'm a pro, everything sucks" performance?

**Comment by rmuir** [comment]
Again, the onus is not on me to prove that your proposed algorithm sucks. thats not how it works.

You have to prove that its better.

We don't need to try to copy what sphinx does. In general, its probably best to do the exact opposite. They do stupid things like removing length normalization completely.

**Comment by brusic** [comment]
You keep avoiding the fact that Lucene already has proximity scoring. IMHO, a far better one.

Sphinx uses LCS, which is a variation of simple edit distance. Lucene uses an exponential function for the decay as the the slop between words increases. Which do you think will be more precise? Look at the second example in the Sphinx documentation. Do you think 1 is the best answer? 

That said, no edit distance method is perfect for every scenario. A query rescorer would be a good place for other methods.

**Comment by kshnurov** [comment]
@rmuir what will make proof for you if not all of this reports saying and proving that term proximity scoring will dramatically improve text search ranking? Basically you're saying "all of this professors and experts are stupid, I know better - we shouldn't do anything".

@brusic yes, Lucene's proximity scoring is even better than LCS, but we can't use it for match queries, as I said above. If we can find a way to use it in conjunction with BM25 without making N! queries - this issue will be solved.

**Comment by rmuir** [comment]
> @rmuir what will make proof for you if not all of this reports saying and proving that term proximity scoring will dramatically improve text search ranking? Basically you're saying "all of this professors and experts are stupid, I know better - we shouldn't do anything"

I don't see dramatic improvements yet anywhere. And the underlying math is still broken and not yet figured out. So yeah, if that's calling people stupid, then i am calling people stupid. Trust me, I am ok with that.

**Comment by kshnurov** [comment]
@rmuir check out this case: https://github.com/elastic/elasticsearch/issues/2867#issuecomment-113305871 Yes, sloppy phrase scoring would make a deal, but isn't it stupid to manually generate N! queries for that?

Can you tell **exactly** what is broken or wrong? You're just making unreasonable statements.

**Comment by rmuir** [comment]
> Can you tell exactly what is broken or wrong?

BM25 formula is based on the assumption that terms are independent. Its as simple as that. We are done here.

**Comment by kshnurov** [comment]
@rmuir what the hell, have you read anything at all? This is **not about BM25**, it's about taking **term proximity** into account! If you don't like BM25, you can use TF/IDF + term proximity.

**Comment by rmuir** [comment]
Which also has the same assumption. I recommend you study where these formulas come from.

**Comment by kshnurov** [comment]
@rmuir are we talking about Elasticsearch? It uses [TF/IDF based scoring](https://www.elastic.co/guide/en/elasticsearch/guide/current/scoring-theory.html), and you're saying it is completely wrong.

**Comment by rmuir** [comment]
I didn't say it was wrong. I said its based on an assumption that terms are mutually independent. Its like this most scoring models. Its fine, it doesn't use proximity in the scoring.

**Comment by kshnurov** [comment]
Ok. ES scoring has a [vector space model](https://www.elastic.co/guide/en/elasticsearch/guide/current/scoring-theory.html#vector-space-model), which (is it?) takes terms dependency into account, but it doesn't include term proximity. So `american elephant` and `american ... (2000 words) ... elephant` are ranked the same, which is wrong. I'm just proposing to add a term proximity on top of it.

**Comment by rmuir** [comment]
You are "just proposing" to add term dependencies to scoring models that assume terms are independent!

**Comment by kshnurov** [comment]
@rmuir yes, so we'll have term dependent scoring model. Maybe not a perfect one, but better than the one we have now in cases [I've described](https://github.com/elastic/elasticsearch/issues/2867#issuecomment-113305871). Is it bad? Impossible? Stupid? If you personally don't need it or like it, it doesn't mean no one needs it. Or you have a better solution, instead of just saying everything is wrong and stupid?

**Comment by rmuir** [comment]
TLDR: we already have a "shitty ad-hoc" proximity scoring, we don't need more of them.

I'm going to explain this to you one last time, so you realize your mistake:

All of the scoring formulas in lucene are based on a bag of words model, they use this model to simplify the world, and based on it use notions like `tf` and `idf`. Its baked into the core of how all these formulas work. 

So I don't want any bogus made-up alterations that Joe Schmoe came up with, that make no mathematical sense, have no real formal basis, and with little or no formal evidence of retrieval improvements.

If you want to do "ad-hoc" scoring like this, then use a rescorer with sloppy phrase query, as we told you already several times. It uses such a "bogus" approach and treats a phrase as a pseudo-term (actually doing the stupid slow work in each document to inefficiently compute true phrase TF) and estimates IDF of a pair of terms by summing up their individual IDFs linearly.

You should also read up a bit more and look at previous experiments: think about your queries and which proximity features you really need, for example, perhaps you only need to consider proximity of neighboring terms, and perhaps only very low slops are useful.  Please, do this instead of complaining about having to use n! phrases, and then whining when I refuse to do your dumb n! for you.

**Comment by kshnurov** [comment]
Low slops won't work if you have 2 matching subsequences with 100 words between them, unless you make a slop query for each subsequence and you'll have to make, let's be precise, `(n-1)!` of them to try all possible subsequences with 2+ words. But who cares, you're always right and all other people are "dumb" "shitty" "idiots" and any text is just a "bag of words" and term proximity is a "bogus approach" you don't need.

That's ok, I got it. "I don't want", "shitty", "bogus", "idiots", "dumb". "best boolean parser on the planet. All boolean parsers in java suck. ... its shittiness ... WTF!!!!" It seems to be your common style. Good luck with that.
@s1monw @clintongormley @kimchy where did you get this guy? He doesn't seem to be adequate at all.

**Comment by clintongormley** [comment]
@kshnurov this feels like an XY question.  You said: "Please implement this algorithm", as opposed to "How can I improve word proximity matching?".  

If you're wanting to take common subsequences into account, I suggest you look at using the shingles token filter to index bigrams.  This combines proximity with TF/IDF in a way that makes sense, and which will perform better at query time than the phrase query.

> where did you get this guy? He doesn't seem to be adequate at all.

Why? Because he disagrees with your point of view? You assume, because there are published papers, that they make sense. @rmuir probably knows more about these algorithms than the rest of Elasticsearch put together. And he spends a lot of time testing them out with real world data. When he has an opinion on these things, I listen to him.  He is also not obliged to spend hours explaining why the maths doesn't make sense.

This is my last post on this topic.
